### PR TITLE
#2001 - Changes to handle PD PPD Part 1

### PR DESCRIPTION
**Pull request by dheepak-aot**
## System changes to handle PD PPD

- [x] Created DB migration to add new type `permanent_disability_status` and add column `pd_status` to student table. Updated the entity model and created enum PDStatus.
- [x] Updated SIMS API, and Vue with PDStatus enum. (By doing this, the PD status field is handled on both student profile and student application)

Student Profile:
image

Student Application:
image

- [x] The existing `pd_verified` column is kept as is to avoid workflow changes as per the story.

- [x] Student profile form definition updated to check for value `Not requested` instead of `Not Requested` to display the `Apply for PD Status` button
- [x] When "apply to PD status" is not allowed due to SIN validation is not complete or already applied for PD status, code changes are made to display the appropriate error message to the toast.

## To be done in upcoming PR
- [ ] Read the PD response from ATBC and update the pd status. Remove/Refactor the existing code.
- [ ] Remove refactor existing code 
- [ ] E2E Tests

### CODE REVIEW

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-pd-status.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+ALTER TABLE
+    sims.students
+ADD
+    COLUMN IF NOT EXISTS pd_status sims.permanent_disability_status;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-pd-status.sql:4]: I hope that it will not open a can of worms about naming possibilities 😉 
I believe that the acronyms for PD/PPD mean
- PPD: Persistent or Prolonged Disability
- PD: Permanent Disability
With the above in mind I believe that instead of `permanent_disability_status` it should be only `disability_status`.

File: sources/packages/backend/apps/api/src/constants/error-code.constants.ts [hunk 2]
```diff
@@ -122,3 +122,8 @@ export const DUPLICATE_SABC_CODE = "DUPLICATE_SABC_CODE";
  * Institution location not valid.
  */
 export const INSTITUTION_LOCATION_NOT_VALID = "INSTITUTION_LOCATION_NOT_VALID";
+
+/**
+ * Request for permanent disability not allowed.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/constants/error-code.constants.ts:127]: The requests are no longer specifically for **permanent** disability.

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts [hunk 3]
```diff
@@ -45,18 +47,17 @@ export class ATBCStudentController extends BaseController {
...
+    // not applied for PD already.
     if (
       !student.sinValidation.isValidSIN ||
-      !!student.studentPDSentAt ||
-      student.studentPDVerified !== null
+      student.pdStatus !== PDStatus.NotRequested
     ) {
       throw new UnprocessableEntityException(
-        "Either the client does not have a validated SIN or the request was already sent to ATBC.",
+        new ApiProcessError(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts:57]: ❤️

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts [hunk 4]
```diff
@@ -45,18 +47,17 @@ export class ATBCStudentController extends BaseController {
...
+          "Either SIN validation is not complete or requested for PD already.",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts:58]: I believe that we can have it adjusted to PD/PPD for now.

File: sources/packages/backend/libs/sims-db/src/entities/student-pd-status.type.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Permanent disability status of student.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-pd-status.type.ts:2]: Everything is SIMS used to be named using PD as permanent disability which is no longer true with the inclusion of PPD.
I would say that we need to reinforce it for the new things and try to adjust the current ones as possible.
As a suggestion, I would remove the "Permanent" from the comment and name the below enum as `DisabilityStatus`.

File: sources/packages/web/src/views/student/StudentProfileEdit.vue [hunk 6]
```diff
@@ -64,9 +65,16 @@ export default defineComponent({
...
+        if (
+          error instanceof ApiProcessError &&
+          error.errorType === PD_REQUEST_NOT_ALLOWED
+        ) {
+          snackBar.error(`${error.message}`);
+          return;
+        }
         snackBar.error(
-          "An error happened during the apply PD process. Please try after sometime.",
+          "Unexpected error while applying for PD. Please try after sometime.",
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentProfileEdit.vue:77]: Can we change PD to PD/PPD?

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts (hunk 4 above)
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts:58]: Is it not better to show separate error message for both the scenarios?

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.institutions.controller.getCompletedApplications.e2e-spec.ts [hunk 7]
```diff
@@ -306,7 +306,7 @@ describe("ApplicationOfferingChangeRequestInstitutionsController(e2e)-getComplet
 
     const searchKeyword = "TestStudent";
     const endpoint = `/institutions/location/${collegeFLocation.id}/application-offering-change-request/completed?page=0&pageLimit=10&searchCriteria=${searchKeyword}`;
-    console.log(endpoint);
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.institutions.controller.getCompletedApplications.e2e-spec.ts:309]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.students.controller.getStudentProfile.e2e-spec.ts [hunk 8]
```diff
@@ -62,7 +62,7 @@ describe("StudentInstitutionsController(e2e)-getStudentProfile", () => {
           },
           phone: student.contactInfo.phone,
         },
-        pdStatus: determinePDStatus(student),
+        pdStatus: student.pdStatus,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.students.controller.getStudentProfile.e2e-spec.ts]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts [hunk 9]
```diff
@@ -45,18 +47,17 @@ export class ATBCStudentController extends BaseController {
     const student = await this.studentService.getStudentById(
       studentUserToken.studentId,
     );
-    // Check the PD status in DB. Student should only be allowed to check the PD status once.
-    // studentPDSentAt is set when student apply for PD status for the first.
-    // studentPDVerified is null before PD scheduled job update status.
-    // studentPDVerified is true if PD confirmed by ATBC or is true from sfas_individual table.
-    // studentPDVerified is false if PD denied by ATBC.
+    // To apply for a PD, SIN validation must be completed for the student and
+    // not applied for PD already.
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts:51]: @dheepak-aot , can you confirm this? In our grooming, I remember saying that the student can apply for PD multiple times, especially for the PPD

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts [hunk 10]
```diff
@@ -45,18 +47,17 @@ export class ATBCStudentController extends BaseController {
...
     if (
       !student.sinValidation.isValidSIN ||
-      !!student.studentPDSentAt ||
-      student.studentPDVerified !== null
+      student.disabilityStatus !== DisabilityStatus.NotRequested
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts:54]: or did u mean only one pending request at a time?

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 11]
```diff
@@ -667,4 +672,19 @@ export class StudentService extends RecordDataModelService<Student> {
...
+   * Get student PD status.
+   * @param pdVerified sfas pdVerified flag.
+   * @returns PD status.
+   */
+  private getPDStatus(pdVerified: boolean): DisabilityStatus {
+    if (pdVerified) {
+      return DisabilityStatus.PD;
+    }
+    if (pdVerified === false) {
+      return DisabilityStatus.Declined;
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:686]: For my info, will we be getting `declined` status from the current ATBC ( I remember we had back-and-forth discussions on this in our grooming)? 
if not, are we only considering the `declined` status only of SFAS entries?

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 12]
```diff
@@ -667,4 +672,19 @@ export class StudentService extends RecordDataModelService<Student> {
       entityManager,
     );
   }
+
+  /**
+   * Get student PD status.
+   * @param pdVerified sfas pdVerified flag.
+   * @returns PD status.
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:679]: Instead of `PD status.`, maybe we can give something as `disability status`, same for line 677, to keep everything generic

File: sources/packages/backend/apps/db-migrations/src/migrations/1689350242512-AddStudentDisabilityStatusCol.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
+import { getSQLFileData } from "../utilities/sqlLoader";
+
+export class AddStudentPDStatusCol1689350242512 implements MigrationInterface {
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/migrations/1689350242512-AddStudentDisabilityStatusCol.ts:4]: class name is not in sync with the file name

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql [hunk 14]
```diff
@@ -0,0 +1,26 @@
+ALTER TABLE
+    sims.students
+ADD
+    COLUMN IF NOT EXISTS disability_status sims.disability_status;
+
+COMMENT ON COLUMN sims.students.disability_status IS 'Disability status of the student.';
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:6]: IMO, we can about the PD and PPD in the comment, it will help the future developer, when they look into the db

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+-- Update the disability_status column based on pd_verified and pd_date_sent for existing data.
+UPDATE
+    sims.students
+SET
+    disability_status = CASE
+        WHEN pd_verified IS NULL
+        AND pd_date_sent IS NOT NULL THEN 'Requested'
+        WHEN pd_verified = true THEN 'PD'
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:15]: Just wondering, In our today's call, you mentioned that we may be removing `pd_verified` in the future, removing it won't affect this code right? As we did, 
image
and this is only for  the existing data in the db right?

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        WHEN pd_verified = false THEN 'Declined'
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:16]: Just wondering, In our today's call, you mentioned that we may be removing `pd_verified` in the future, removing it won't affect this code right? As we did, 
image
and this is only for the existing data in the db, right?

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        ELSE 'Not requested' :: sims.disability_status
+    END;
+
+-- Set not null constraint after updating the column.
+ALTER TABLE
+    sims.students
+ALTER COLUMN
+    disability_status
+SET
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:25]: since, its not null, how about adding the default as `NotRequested`?

File: sources/packages/backend/libs/integrations/src/atbc-integration/atbc-integration.processing.service.ts [hunk 18]
```diff
@@ -45,7 +45,7 @@ export class ATBCIntegrationProcessingService {
     if (response) {
       const auditUser = await this.systemUsersService.systemUser();
       // Update PD sent date.
-      await this.studentService.updatePDSentDate(student.id, auditUser);
+      await this.studentService.updatePDRequested(student.id, auditUser);
```
Comment (ann-aot) [inline sources/packages/backend/libs/integrations/src/atbc-integration/atbc-integration.processing.service.ts:48]: since you already updated the method name, maybe you can use `disability status` here too instead of `pd`

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-pd-status.sql (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-pd-status.sql]: As per discussion with the dev team, we decided to use disability status instead of permanent disability status and try to change the existing code as much as possible.

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 19]
```diff
@@ -166,11 +168,14 @@ export class StudentService extends RecordDataModelService<Student> {
...
-        studentSIN,
-      );
+      const sfasStudentPDVerified =
+        await this.sfasIndividualService.getPDStatus(
+          user.lastName,
+          student.birthDate,
+          studentSIN,
+        );
+      student.studentPDVerified = sfasStudentPDVerified;
+      student.disabilityStatus = this.getPDStatus(sfasStudentPDVerified);
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:178]: When the student is not present in the SFAS system, then also a PD Status will be returned, right? However, in that scenario we actually do not know the PD Status.

Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:178]: Maybe not part of the PR, but, `getPDStatus` will return `undefined` even if the student is not present in SFAS, then does this `try catch` matter?

for that scenario(new student), is the plan to add `not requested` from line 178 and do we need line 177?

File: sources/packages/backend/libs/sims-db/src/entities/student-disability-status.type.ts [hunk 20]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Disability status of student.
+ */
+export enum DisabilityStatus {
+  NotRequested = "Not requested",
+  Requested = "Requested",
+  PD = "PD",
+  PPD = "PPD",
```
Comment (ann-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-disability-status.type.ts:8]: IMO, maybe adding a brief comments to each status maybe help the future developers

File: sources/packages/backend/apps/api/src/services/student/student.service.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:686]: As of now, we are considering declined only for SFAS entries.

File: sources/packages/backend/libs/sims-db/src/entities/student.model.ts [hunk 21]
```diff
@@ -101,4 +101,15 @@ export class Student extends RecordDataModel {
...
   studentRestrictions?: StudentRestriction[];
+
+  /**
+   * Disability status of the student.
+   */
+  @Column({
+    name: "disability_status",
+    type: "enum",
+    enum: DisabilityStatus,
+    enumName: "PDStatus",
```
Comment (ann-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student.model.ts:112]: why enum and enum name are different?

File: sources/packages/backend/libs/test-utils/src/factories/student.ts [hunk 22]
```diff
@@ -23,6 +23,7 @@ export function createFakeStudent(user?: User): Student {
     phone: faker.phone.phoneNumber(),
   };
   student.sinConsent = true;
+  student.disabilityStatus = DisabilityStatus.NotRequested;
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/student.ts:26]: addressing https://github.com/bcgov/SIMS/pull/2115/files#r1265791873 can eliminate this

File: sources/packages/backend/apps/api/src/services/student/student.service.ts (hunk 11 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:686]: Actually, addressing https://github.com/bcgov/SIMS/pull/2115/files#r1265791873 can eliminate this can avoid adding status to a new student

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql (hunk 16 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:16]: yes @ann-aot. This is only for existing data. (mainly keeping in mind test environment).

File: sources/packages/web/src/services/StudentService.ts [hunk 23]
```diff
@@ -73,8 +73,8 @@ export class StudentService {
     await ApiClient.Students.synchronizeFromUserToken();
   }
 
-  async applyForPDStatus(): Promise<void> {
-    return ApiClient.Students.applyForPDStatus();
+  async applyForDisabilityStatus(): Promise<void> {
+    return ApiClient.Students.applyForDisabilityStatus();
```
Comment (ann-aot) [inline sources/packages/web/src/services/StudentService.ts:77]: not part of the PR, but do we need `return` here?

File: sources/packages/web/src/types/contracts/StudentContract.ts [hunk 24]
```diff
@@ -10,13 +10,14 @@ export interface StudentProfile extends StudentProfileAPIOutDTO {
...
 /**
- * Enumeration for student permanent disability status.
+ * Permanent disability status of student.
  */
-export enum StudentPDStatus {
-  Yes = "Yes",
-  No = "No",
-  NotRequested = "Not Requested",
-  Pending = "Pending",
+export enum PDStatus {
```
Comment (ann-aot) [inline sources/packages/web/src/types/contracts/StudentContract.ts:15]: this name can also be renamed

**Review by ann-aot — COMMENTED** [review]
Added some comments

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql (hunk 17 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:25]: I had the same thought initially. But because due to the reason that we manually set a default value from the code, I preferred not to have a default at SQL as well. 

image

image

File: sources/packages/backend/apps/api/src/services/student/student.service.ts (hunk 19 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:178]: https://github.com/bcgov/SIMS/pull/2115/files#r1265791873 this can avoid adding status to a new student

Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:178]: If the student is not present in SFAS, getPDStatus() (refacored to getDisabilityStatus) will return `Not required` 

image

The existing try catch I believe is to ensure that, we should not go further to create a student if we cannot get the details from SFAS table as it could end  up creating student with wrong disability status.

Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts]: The intention of the logic is, if the newly created student is already present is SFAS, then read the pd_status, otherwise create the student with default PD status which is "Not requested"

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts]: Thanks for bring this up. 
I had a quick chat with @JasonCTang on the same and got to know that, as per the current understanding, student requests disability status only once in our application. In case they want to apply again because of something happened to them later after losing their status, they update ATBC directly and ATBC will be updating the student disability status which we read on daily basis through their response API.
I will update this information to the ticket.

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql (hunk 17 above)
Comment (guru-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:25]: I feel, its good this way, rather setting it as a default value in the SQL.

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+-- Update the disability_status column based on pd_verified and pd_date_sent for existing data.
+UPDATE
+    sims.students
+SET
+    disability_status = CASE
+        WHEN pd_verified IS NULL
+        AND pd_date_sent IS NOT NULL THEN 'Requested'
+        WHEN pd_verified = true THEN 'PD'
+        WHEN pd_verified = false THEN 'Declined'
+        ELSE 'Not requested' :: sims.disability_status
```
Comment (guru-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:17]: I get this update is written for the existing data in Test, but is there a chance that any PDs can be PPDs? are we considering any update in the existing data update from ATBC?

File: sources/packages/backend/libs/integrations/src/services/student/student.service.ts [hunk 26]
```diff
@@ -32,23 +33,25 @@ export class StudentService extends RecordDataModelService<Student> {
...
-      studentToUpdate.modifier = auditUser;
-      studentToUpdate.updatedAt = now;
-      return this.repo.save(studentToUpdate);
-    }
+  async updateDisabilityRequested(
+    studentId: number,
+    auditUser: User,
+  ): Promise<UpdateResult> {
+    const now = new Date();
+    return this.repo.update(
```
Comment (guru-aot) [inline sources/packages/backend/libs/integrations/src/services/student/student.service.ts:46]: ❤️

File: sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql (hunk 25 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Student/Add-col-disability-status.sql:17]: There is no PPD in our system so far so we cannot do anything in the query about it.

If a student is PPD, ATBC response will be read and updated in our system.

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts]: @sh16011993 Sorry I do not follow. What are the error messages are you talking about?

Comment (sh16011993) [reply sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts]: Resolving it as per the conversation with @dheepak-aot

**Review by sh16011993 — APPROVED** [review]
LGTM @dheepak-aot 👍

File: sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts (hunk 9 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/route-controllers/atbc/atbc.students.controller.ts]: Thanks @dheepak-aot for the details. Ya, adding it to the ticket will make it easier for future development

File: sources/packages/web/src/services/http/dto/Student.dto.ts [hunk 27]
```diff
@@ -15,6 +15,8 @@ export class CreateStudentAPIInDTO extends AddressDetailsFormAPIDTO {
   sinNumber: string;
   @Expose()
   gender: string;
+  @Expose()
+  sinConsent: boolean;
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/dto/Student.dto.ts:19]: ❤️

**Review by ann-aot — APPROVED** [review]
Nice work @dheepak-aot. thanks for doing the changes, wave fix as well as the label changes 👍

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work @dheepak-aot
